notebooks/ima_viewer/ima_tools.py

```diff
@@ -0,0 +1,747 @@
+"""
+    This script contains functions written for the IMA
+    Viewer notebook in the WFC3Library repo.
+
+    Author
+    ------
+    Mariarosa Marinelli, Dec 2022
+
+    Use
+    ---
+    The functions in this script are intended to be
+    imported, i.e.:
+
+        from ima_tools import make_interactive_plot
+"""
+
+from astropy.io import fits
+from astroquery.mast import Observations
+import imageio.v2 as imageio
+from IPython.display import display
+from ipywidgets import interact, interactive, Layout
+import ipywidgets as widgets
+import matplotlib
+import matplotlib.pyplot as plt
+import matplotlib.ticker as tick
+from matplotlib.colors import Normalize
+import numpy as np
+import os
+import shutil
+
+# Plotting functions:
+def get_data_and_norm_lims(ext, fits_file, pct_limits=(50, 99.5)):
+    """
+    Helper function to grab the data from a specific
+    extension, find the normalization factors from the
+    percentage limits either set by the user or from the
+    default percentages.
+
+    Parameters
+    ----------
+    ext : int
+        Integer corresponding to the extension of the
+        fits file.
+    fits_file : fits
+        Fits file at hand.
+    pct_limits : tuple of float
+        Percentage limits as set by user. Default is a
+        lower bound of 50% and upper bound of 99.5%.
+
+    Returns
+    -------
+    data : array
+        Array of data from the specified extension.
+    limits : tuple of float
+        Lower and upper limits, specific to the extension.
+    """
+    data = fits_file[ext].data
+    data_min = np.nanpercentile(data, pct_limits[0])
+    data_max = np.nanpercentile(data, pct_limits[1])
+    limits = (data_min, data_max)
+    return data, limits
+
+def get_indices(fits_file, ext_type):
+    """
+    Helper function to grab the indices that are
+    relevant and match the extension type specified.
+
+    Parameters
+    ----------
+    fits_file : fits
+        Fits file at hand.
+    ext_type : str
+        Extension type.
+
+    Returns
+    -------
+    indices : list of int
+        List of extensions, ordered by time.
+    """
+    indices = [i for i, ext in enumerate(fits_file) if ext.name == ext_type]
+    indices.reverse()
+    return indices
+
+def make_interactive_plot(filepath, ext_type='SCI'):
+    """
```

Comment:
Thoughts for version 2.0 or version 3.0:

An absolute scale slider would be nice too. Since it's a percentage scaling, the color map vmin/vmax values changes slightly from extension to extension. An absolute scale (i.e., from some min value in electrons to some max value in electrons) would make it easier to directly compare between extensions since the color map would be constant.

DQ arrays in this framework don't represent the data as well as it could be. The percentage scaling is somewhat meaningless in this case. Maybe do a discrete map just showing one flag? For example, add parameter `discrete_slider`. If True, the slider will choose one value instead of a range. If False, use the range of values (this combined with the absolute scaling could be neat). Open to chatting/white boarding more about this.